src/control/lib/daos/api/provider.go

```diff
@@ -0,0 +1,48 @@
+//
+// (C) Copyright 2024 Intel Corporation.
+//
+// SPDX-License-Identifier: BSD-2-Clause-Patent
+//
+
+package api
+
+import (
+	"github.com/pkg/errors"
+
+	"github.com/daos-stack/daos/src/control/logging"
+)
+
+type (
+	debugTraceLogger interface {
+		logging.TraceLogger
+		logging.DebugLogger
+	}
+
+	// Provider wraps the C DAOS API with methods that accept and return
+	// standard Go data types.
+	Provider struct {
+		log     debugTraceLogger
+		api     *api
+		cleanup func()
+	}
+)
+
+// NewProvider returns an initialized DAOS API provider.
+func NewProvider(log debugTraceLogger) (*Provider, error) {
+	api := &api{}
+	cleanup, err := api.Init()
```

Comment:
I find myself wondering if the `api` struct is an unnecessary layer. It doesn't seem that other API calls are going through it.

Reply:
Yeah... I wasn't planning on adding the API stuff yet, but I realized that I needed the sub-package in order to move the `GetSystemInfo()` implementation. We don't want the `daos` package to have a linker dependency on `libdaos.so` because it provides types that are used by the server too.

Longer term, I think that I want the `api` type to provide a mocking point to allow for easier unit testing, but tbh I still haven't settled on an approach that I'm completely happy with.

Reply:
I think we should try to be consistent across the different bindings on how we handle API mocking, we have a number of different approaches and it would be good to settle on one. I'm biased but I liked the way the pciutils bindings ended up linking in with Kris' hardware library framework.